pybind/cephfs, mgr/volumes: introduce non-recurisve rmtree(), refactor purge() to use it and add MDS optimizations

rishabh-d-dave opened the pull request:
Method purge() in trash.py calls rmtree() which is recursive method. To
avoid Python's recurision limit, switch to non-recursive approach.

Fixes: https://tracker.ceph.com/issues/71648

Also fixes https://tracker.ceph.com/issues/72779 and https://tracker.ceph.com/issues/72993.






<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)
</details>

rishabh-d-dave commented on src/pybind/mgr/volumes/fs/operations/trash.py:
```diff
@@ -59,31 +62,27 @@ def purge(self, trashpath, should_cancel):
+                    file_list = listdir(self.fs, x, filter_files=False)
+                    if file_list:
+                        for i in file_list:
+                            stack.append(os.path.join(x, i))
```
I agree, I was pondering on the same -- `listdir()` vs `readdir()`. `listdir()` fetches and stores all dirents making it BFS and storing all dirents at once increases the chance of running out of memory. Switching to `readdir()` should fix it.

rishabh-d-dave replied:
Done.

rishabh-d-dave commented:
jenkins test make check arm64

batrick commented on src/pybind/mgr/volumes/fs/operations/trash.py:
```diff
@@ -53,37 +55,75 @@ def get_trash_entry(self, exclude_list):
+            # repetitive calls to fs.opendir().
+            stack = deque([(b'', self.fs.opendir(trashpath)), ])
+            while stack and not should_cancel():
+                log.info(f'here123 1')
```
```suggestion
```

batrick commented on src/pybind/mgr/volumes/fs/operations/trash.py:
```diff
@@ -53,37 +55,75 @@ def get_trash_entry(self, exclude_list):
+                    # stack and then deleting it and popping it from stack,
+                    # since it can be deleted right away.
+                    if de.is_dir():
+                        if is_dir_empty(self.fs, de_path):
```
this call to `is_dir_empty` should be avoided: keep everything in this method.

rishabh-d-dave replied:
That would make this method too bloated IMO. Therefore, I've added a `is_dir_empty()` within this method which doesn't use any helper method and this method is being called instead.

batrick commented on src/pybind/mgr/volumes/fs/operations/trash.py:
```diff
@@ -53,37 +55,75 @@ def get_trash_entry(self, exclude_list):
+            #
+            # The directory handle is also stored on stack to avoid unnecessary
+            # repetitive calls to fs.opendir().
+            stack = deque([(b'', self.fs.opendir(trashpath)), ])
```
```suggestion
            stack = deque([(b'', self.fs.opendir(trashpath)), ])
```

a list is fine

rishabh-d-dave replied:
You mean a list instead of tuple? Or deque? Both tuple and deque are perform better in this scenario.

batrick commented on src/pybind/mgr/volumes/fs/operations/trash.py:
```diff
@@ -53,37 +55,75 @@ def get_trash_entry(self, exclude_list):
+                            de_rel_path = de_path.replace(trashpath, b'')
+                            stack.append((de_rel_path, self.fs.opendir(de_path)))
+                            break
+                    elif de.is_file():
+                        self.fs.unlink(de_path)
```
```suggestion
                    else:
                        self.fs.unlink(de_path)
```

rishabh-d-dave replied:
Done.

batrick commented on src/pybind/mgr/volumes/fs/operations/trash.py:
```diff
@@ -53,37 +55,75 @@ def get_trash_entry(self, exclude_list):
+
+                if is_dir_empty(self.fs, curr_dir_path):
+                    self.fs.rmdir(curr_dir_path)
+                    stack.pop()
```
ah, I thought this was outside somewhere. You can just put this at the below this loop, no?

rishabh-d-dave replied:
I don't get your comment. It's inside the outer loop and moving it to inner loop won't work.

batrick commented on src/pybind/mgr/volumes/fs/operations/trash.py:
```diff
@@ -53,37 +54,89 @@ def get_trash_entry(self, exclude_list):
+
+                    de = self.fs.readdir(curr_dir_handle)
+
+                if is_dir_empty(self.fs, curr_dir_path):
+                    self.fs.rmdir(curr_dir_path)
+                    stack.pop()
```
Also, let the parent directory actually try the call to `rmdir` to keep the removal at the parent level of the stack.

batrick replied:
Note: an "empty" directory with snapshots will cause this to endlessly recurse. That shouldn't happen with a subvolume but your code should still be resilient to it.

I think the right thing to do is to keep track of errors (e.g. cannot opendir or unlink or rmdir). Try to unlink / recursively descend, and then skip any dentries present in a dict you maintain at each level in the stack. IOW, when a child directory cannot be completely removed, the parent directory should remember that and not retry for that child directory.

rishabh-d-dave commented on src/pybind/mgr/volumes/fs/operations/trash.py:
```diff
@@ -53,37 +54,89 @@ def get_trash_entry(self, exclude_list):
+                    # stack and then deleting it and popping it from stack,
+                    # since it can be deleted right away.
+                    if de.is_dir():
+                        if is_dir_empty(self.fs, de_path):
+                            self.fs.rmdir(de_path)
+                        else:
+                            de_rel_path = de_path.replace(trashpath, b'')
+                            stack.append((de_rel_path, self.fs.opendir(de_path)))
+                            break
```
Much better! Thank you for the suggestion!

rishabh-d-dave commented on src/pybind/mgr/volumes/fs/operations/trash.py:
```diff
@@ -53,37 +54,89 @@ def get_trash_entry(self, exclude_list):
-                self.fs.rmdir(root_path)
-
-        # catch any unlink errors
+        def is_dir_empty(fs, dirpath):
+            '''
+            Return true if the directory in given path is empty.
+            '''
+            # number of directory entries (excluding "." and "..").
+            dirent_count = 0
+
+            with self.fs.opendir(dirpath) as dir_handle:
+                de = self.fs.readdir(dir_handle)
+                while de:
+                    if de.d_name not in (b'.', b'..'):
+                        dirent_count += 1
+                    de = self.fs.readdir(dir_handle)
+
+            return False if dirent_count else True
```
Right.

rishabh-d-dave commented on src/pybind/mgr/volumes/fs/operations/trash.py:
```diff
@@ -53,37 +54,89 @@ def get_trash_entry(self, exclude_list):
+                curr_dir_path = os.path.join(trashpath, stack[len(stack) - 1][0])
+                curr_dir_handle = stack[len(stack) - 1][1]
+
+                # de = directory entry
```
Done.

rishabh-d-dave commented on src/pybind/mgr/volumes/fs/operations/trash.py:
```diff
@@ -53,37 +54,81 @@ def get_trash_entry(self, exclude_list):
+                    de_path = os.path.join(curr_dir_path, de.d_name)
+                    # optimization: if directory is empty, avoid pushing it on
+                    # stack and then deleting it and popping it from stack,
+                    # since it can be deleted right away.
```
Removed.

batrick reviewed (changes requested):
Add more error handling and refactor the code a bit.

rishabh-d-dave commented:
@batrick I had some confusion regarding in the code suggested in last review cycle but I think I've got it sorted. I've also made some improvements along the way. PTAL, I'll mark previous review threads resolved based on next review cycle.

rishabh-d-dave commented:
jenkins test make check arm64

rishabh-d-dave commented:
jenkins test make check arm64

batrick commented on src/pybind/mgr/volumes/fs/operations/trash.py:
```diff
@@ -53,37 +54,139 @@ def get_trash_entry(self, exclude_list):
+        except cephfs.ObjectNotFound:
+            return
         except cephfs.Error as e:
             raise VolumeException(-e.args[0], e.args[1])
```
who gets this exception?

batrick commented on src/pybind/mgr/volumes/fs/operations/trash.py:
```diff
@@ -53,37 +54,139 @@ def get_trash_entry(self, exclude_list):
+                    curr_dir.handle.rewinddir()
+
+            self.fs.rmdir(trashpath)
+        except cephfs.ObjectNotFound:
```
`ObjectNotEmpty`?

rishabh-d-dave replied:
> In fact, if you unlink and get that error then I would consider that success, no?

I think so too but this exception was already present in earlier version, so I wasn't sure if it should if removed. And if it should, whether it should be removed on this PR.

rishabh-d-dave replied:
> I would say you should ignore it in every place we rmdir/unlink.

Right now we are doing that since we simply return if that exception occurs -

```
        except cephfs.ObjectNotFound:
            return
```

rishabh-d-dave commented on src/pybind/mgr/volumes/fs/operations/trash.py:
```diff
@@ -53,37 +54,139 @@ def get_trash_entry(self, exclude_list):
+
+                    is_curr_dir_empty = False
+                    if de.d_name in curr_dir.de_ignore_list:
+                        continue
```
Done.

rishabh-d-dave commented on src/pybind/mgr/volumes/fs/operations/trash.py:
```diff
@@ -53,37 +54,159 @@ def get_trash_entry(self, exclude_list):
+                        if de.d_name in (b'.', b'..'):
+                            de = self.fs.readdir(self.handle)
+                        elif de.d_name in curr_dir.de_ignore_list:
+                            self.is_empty = False
```
Marking this resolved, this is not needed due to this change - https://github.com/ceph/ceph/pull/63917#discussion_r2164050279

rishabh-d-dave commented on src/pybind/mgr/volumes/fs/operations/trash.py:
```diff
@@ -53,37 +54,159 @@ def get_trash_entry(self, exclude_list):
+                            pass
+
+                    stack.pop()
+                else:
```
Fixed.

rishabh-d-dave commented:
@batrick I've moved rewinddir() from outer loop to inner loop since rewinddir() needs to be called when readdir() and unlink()/rmdir() calls are interleaved.

rishabh-d-dave commented:
jenkins test api

rishabh-d-dave commented:
jenkins test make check

rishabh-d-dave commented:
jenkins retest this please

rishabh-d-dave commented:
.

rishabh-d-dave commented:
.

rishabh-d-dave commented on src/pybind/mgr/volumes/fs/operations/trash.py:
```diff
@@ -53,37 +54,168 @@ def get_trash_entry(self, exclude_list):
+                return d_name in self.de_ignore_list
+
+            def has_any_fs_op_failed(self):
+                return self.has_readdir_failed or len(self.de_ignore_list) > 1
```
Fixed.

rishabh-d-dave commented on src/pybind/mgr/volumes/fs/operations/trash.py:
```diff
@@ -53,37 +54,168 @@ def get_trash_entry(self, exclude_list):
+        # unexpected results.
+        #
+        # os.path.join('abcd', '123') = '/abcd/123'   --> expected
+        # os.path.join('abcd', '/123') = '/123'       --> unexpected
```
This block of code is not necesssary anymore. I've removed it.

batrick commented on src/pybind/cephfs/cephfs.pyx:
```diff
@@ -2832,3 +2837,148 @@ cdef class LibCephFS(object):
+                        continue
+
+                if de_has_been_removed:
+                    curr_dir.handle.rewinddir()
```
Only do this after completing iteration. `unlink` does not invalidate the previously fetched contents of `getdents` for `readdir`.

The important point here is that `unlink` may cause the directory contents to reorder.

batrick commented on src/test/pybind/test_cephfs.py:
```diff
@@ -845,6 +845,66 @@ def test_empty_snapshot_info(testdir):
+    '''
+    Test rmtree() with different kinds of file hierarchies.
+    '''
+    should_cancel = lambda: False
```
don't forget to test `should_cancel`.

batrick commented on src/test/pybind/test_cephfs.py:
```diff
@@ -845,6 +845,66 @@ def test_empty_snapshot_info(testdir):
+        fd = cephfs.open(f'/dir1/file{i}', 'w', 0o755)
+        cephfs.write(fd, b'abcd', 0)
+        cephfs.close(fd)
+    cephfs.rmtree('dir1', should_cancel)
```
separate tests for each of these.

batrick replied:
Verify that `stat` on `'diri'` fails with `ENOENT`.

batrick commented on src/test/pybind/test_cephfs.py:
```diff
@@ -845,6 +845,66 @@ def test_empty_snapshot_info(testdir):
+        fd = cephfs.open(f'/dir1/file{i}', 'w', 0o755)
+        cephfs.write(fd, b'abcd', 0)
+        cephfs.close(fd)
+    cephfs.chmod('/dir1/file1', 0o000)
```
```suggestion
    cephfs.chmod('/dir1/', 0o000)
```

batrick replied:
`rm` will complain that `/dir1/file1` is write protected but that does not prevent removal.

batrick commented on src/test/pybind/test_cephfs.py:
```diff
@@ -845,6 +845,66 @@ def test_empty_snapshot_info(testdir):
+        cephfs.close(fd)
+    cephfs.chmod('/dir1/file1', 0o000)
+    cephfs.rmtree('dir1', should_cancel)
+
```
Test with a snapshot on `dir1` too.

batrick commented on src/pybind/cephfs/cephfs.pyx:
```diff
@@ -2832,3 +2837,184 @@ cdef class LibCephFS(object):
+
+            def __str__(self):
+                return self.path
+
```
```suggestion
     def unlink(...)
          self.write_epoch = self.read_epoch
```

batrick commented on src/pybind/cephfs/cephfs.pyx:
```diff
@@ -2832,3 +2837,184 @@ cdef class LibCephFS(object):
+                            de = self.fs.readdir(self.handle)
+                        else:
+                            self.is_empty = False
+                            return de
```
```suggestion
                            return de
   if self.write_epoch == self.read_epoch:
        self.handle.rewinddir()
        self.read_epoch += 1
        return self.read_dir()
   
```

batrick commented on src/pybind/cephfs/cephfs.pyx:
```diff
@@ -2832,3 +2837,184 @@ cdef class LibCephFS(object):
+                        try:
+                            stack.append(Dir(de_path, self))
+
+                            # Call rewinddir() or else future readdir() for this
+                            # dir will have undefined behaviour.
+                            curr_dir.handle.rewinddir()
```
```suggestion

```

rishabh-d-dave commented:
@batrick The approach for checking the return value of readdir() seemed simpler, therefore I made this change instead of adding epochs.

rishabh-d-dave commented:
jenkins test make check

batrick commented on src/test/pybind/test_cephfs.py:
```diff
@@ -940,3 +940,246 @@ def test_multi_target_command():
+
+            file_name = f'/dir3/file{i}'.encode('utf-8')
+            slink_name = f'slink{i}'.encode('utf-8')
+            cephfs.link(file_name, slink_name)
```
```suggestion
            cephfs.symlink(file_name, slink_name)
```

?

batrick commented on src/test/pybind/test_cephfs.py:
```diff
@@ -940,3 +940,246 @@ def test_multi_target_command():
+
+        cephfs.rmtree('dir4', should_cancel)
+        assert_raises(libcephfs.ObjectNotFound, cephfs.stat, 'dir4')
+
```
add test for tree with symlinks referring to parent e.g. "a -> ../"

batrick commented on src/test/pybind/test_cephfs.py:
```diff
@@ -940,3 +940,246 @@ def test_multi_target_command():
+            cephfs.close(fd)
+
+        cephfs.chmod('/dir1', 0o000)
+        cephfs.rmtree('dir1', should_cancel)
```
This doesn't throw an exception?

rishabh-d-dave replied:
We have a separate test now to cover this case (since we can switch on and off whether exception is raised).

batrick commented on src/test/pybind/test_cephfs.py:
```diff
@@ -940,3 +940,246 @@ def test_multi_target_command():
+            cephfs.close(fd)
+        cephfs.mksnap('/dir1/dir2', 'snap1', 0o755)
+
+        cephfs.rmtree('dir1', should_cancel)
```
ditto

batrick commented on src/pybind/cephfs/cephfs.pyx:
```diff
@@ -2832,3 +2837,186 @@ cdef class LibCephFS(object):
            free(buf)
+
+
+    def rmtree(self, trash_path, should_cancel):
```
I would suggest two modes for this:

`suppress_errors = False` (default False)

The ceph-mgr will use `suppress_errors = True` to try to get as far as possible. There should be a return value indicating whether any errors were encountered (perhaps, a count instead of a specific error).

If it is `False`, then immediately raise an exception and quit `rmtree`.

batrick commented on src/test/pybind/test_cephfs.py:
```diff
@@ -940,3 +940,246 @@ def test_multi_target_command():
+        Test that rmtree() stops deleting the file hierarchy when the return
+        value of "should_cancel" becomes True.
+        '''
+        cancel_flag = False
```
```suggestion
        cancel_flag = threading.Event()
```

batrick commented on src/test/pybind/test_cephfs.py:
```diff
@@ -940,3 +940,246 @@ def test_multi_target_command():
+        value of "should_cancel" becomes True.
+        '''
+        cancel_flag = False
+        should_cancel = lambda: cancel_flag
```
To make this not racy, I suggest adding a sleep to this `should_cancel` procedure such that it would take more than 0.5 seconds to finish unlinking dir6.

rishabh-d-dave commented on src/test/pybind/test_cephfs.py:
```diff
@@ -940,3 +940,246 @@ def test_multi_target_command():
+
+        cephfs.rmtree('dir5', should_cancel)
+        assert_raises(libcephfs.ObjectNotFound, cephfs.stat, 'dir5')
+
```
Done.

rishabh-d-dave commented on src/pybind/cephfs/cephfs.pyx:
```diff
@@ -2832,3 +2837,186 @@ cdef class LibCephFS(object):
+                except Error as e:
+                    log.error('Exception occured during opendir(): '
+                              f'"{e}"')
+                    self.add_to_de_ignore_list(de_name)
```
It should've been added to parent's de_ignore_list. Done.

rishabh-d-dave commented:
@batrick PTAL. I've made the changes as per our yeseterday's discussion.

rishabh-d-dave commented:
@batrick I've made the changes and moved cptree() code to a separate PR. The helper and the rmtree() method were passing around a set of variables around all the time. Also they stood out as a separate group within LibCephFS and we have to prefix their name with `_` and `__` . I've moved them in a class, it takes away all this trouble, makes the code more clean and more compact too.

PTAL and let me know if it looks good, I intend to merge third commit into first commit.

batrick commented on src/pybind/cephfs/cephfs.pyx:
```diff
@@ -199,6 +205,14 @@ class DiskQuotaExceeded(OSError):
 
+class OpCancelled(OSError):
+    def __init__(self, op_name):
+        self.errno = 125
```
Why didn't you define this the same way as the others?

Also there's a bug: https://tracker.ceph.com/issues/72401

Please resolve it in a followup PR.

batrick commented on src/pybind/cephfs/cephfs.pyx:
```diff
@@ -2832,3 +2846,288 @@ cdef class LibCephFS(object):
+    def has_any_fs_op_failed(self):
+        return self.has_readdir_failed or len(self.de_ignore_list) > 0
+
+    def try_opendir(self):
+        try:
+            return self.fs.opendir(self.path)
+        except Error as e:
+            log.error('Exception occured during opendir(): '
+                      f'"{e}"')
+            raise
```
I don't really see a reason for this helper to exist.

batrick commented on src/pybind/cephfs/cephfs.pyx:
```diff
@@ -2832,3 +2846,288 @@ cdef class LibCephFS(object):
+            else:
+                raise
+
+        # New dir has been added to the stack, call rewinddir() since the new
+        # dir will be traversed now. Else, future call to current dir's
+        # readdir() will have implementation-defined behaviour, which might be
+        # different from the expected behaviour.
+        self.curr_dir.handle.rewinddir()
```
should not be necessary.

batrick commented on src/pybind/cephfs/cephfs.pyx:
```diff
@@ -2832,3 +2846,288 @@ cdef class LibCephFS(object):
+                        retval = self.curr_dir.try_rmdir(de_name,
+                                                         self.suppress_errors)
+                    except ObjectNotEmpty:
+                        retval = self.add_dir_to_stack(de_name)
```
If this returns `"break"` then it will stop processing the directory, right? That should not be correct.

I think you can structure this so that this `break` return is not necessary. You already removed it for `try_unlink`.

rishabh-d-dave replied:
> If this returns "break" then it will stop processing the directory, right? That should not be correct.

Done.

> I think you can structure this so that this break return is not necessary. You already removed it for try_unlink.

We are not returning break anymore but we'll need to indicate whether adding a new dir to stack failed or succeeded, based on which the iteration in the inner loop can be halted or allowed to continue. We are doing that now by returning boolean value which is an improvement in comparison.

rishabh-d-dave commented:
I've been running the unit tests with teuthology over last couple of weeks. All of the issues have been fixed and tests are running successfully since last few times -

https://pulpito.ceph.com/rishabh-2025-08-06_10:55:53-fs:libcephfs-rishabh-rmtree-17-testing-default-smithi/

batrick commented on qa/workunits/fs/test_python.sh:
```diff
@@ -1,6 +1,8 @@
+username=$(id -un)
+sudo chown -R $username:$username $(dirname $0)/../../../../
+
+python3 -m pytest -v $(dirname $0)/../../../src/test/pybind/test_cephfs.py
```
What? How is this failing if you run the python script without root?

batrick commented on src/pybind/cephfs/cephfs.pyx:
```diff
@@ -2832,3 +2849,284 @@ cdef class LibCephFS(object):
+                        retval = self.curr_dir.try_rmdir(de_name,
+                                                         self.suppress_errors)
+                    except ObjectNotEmpty:
+                        retval = self.add_dir_to_stack(de_name)
+                        if retval:
```
```suggestion
                        if self.add_dir_to_stack(de_name):
```

batrick commented on src/pybind/cephfs/cephfs.pyx:
```diff
@@ -2832,3 +2849,284 @@ cdef class LibCephFS(object):
+                de_name = de.d_name.decode()
+                if de.is_dir():
+                    try:
+                        retval = self.curr_dir.try_rmdir(de_name,
```
```suggestion
                        self.curr_dir.try_rmdir(de_name,
```

batrick commented on src/pybind/cephfs/cephfs.pyx:
```diff
@@ -2832,3 +2849,284 @@ cdef class LibCephFS(object):
+                            # stack.
+                            break
+                else:
+                    retval = self.curr_dir.try_unlink(de_name,
```
```suggestion
                    self.curr_dir.try_unlink(de_name,
```

rishabh-d-dave commented on src/pybind/cephfs/cephfs.pyx:
```diff
@@ -2832,3 +2849,284 @@ cdef class LibCephFS(object):
+                    self.is_empty = False
+                    return de
+
+                de = self.fs.readdir(self.handle)
+                if not de and self.de_has_been_removed:
+                    log.debug('rewinding and restarting reading current dir '
+                              f'"{self.name}", since a dentry has been '
+                              'removed.')
+                    self.handle.rewinddir()
+                    # reset de_has_been_removed flag
+                    self.de_has_been_removed = False
+                    de = self.fs.readdir(self.handle)
```
Moving this block outside `while` loop will cause it to iterate infinitely since then there'll be no code to update value of `de` .

batrick replied:
I meant to keep `de = self.fs.readdir(self.handle)` in the while loop.

rishabh-d-dave commented:
https://jenkins.ceph.com/job/ceph-windows-pull-requests/66628/

rishabh-d-dave commented:
@batrick ping

rishabh-d-dave commented on src/mds/CInode.cc:
```diff
@@ -1096,14 +1096,37 @@ bool CInode::is_projected_ancestor_of(const CInode *other) const
+   * useful but also it makes reading logs harder. Therefore, shorten the path
+   * when used for logging.
+   */
+  static int path_comp_count = 1;
```
Okay. It's almost done, might require a normal change. Will look into it tomorrow. Feel free to review/approve in the mean time.

batrick reviewed (changes requested):
nearly there!

rishabh-d-dave commented:
jenkins test make check

batrick commented on src/mds/Server.cc:
```diff
@@ -10205,8 +10221,24 @@ void Server::_rename_prepare(const MDRequestRef& mdr,
+	}
+	if (trimmed_t == "" && t != "") {
+	  trimmed_t = "..." + t.substr(n, -1);
+	}
```
Please put this in one of the src/common/str* helper libraries so you're not repeating the code.

batrick reviewed (approved):
Otherwise, LGTM. Remaining nits can be addressed with filepath refactor.
